### DISCUSSION TITLE
FRAX-like USD peg

### DISCUSSION THREAD

**Discussion by rndquu** [pavlovcik/uad-ui#829] [Q&A]
We have plans of putting "on hold" [Credit](https://github.com/ubiquity/ubiquity-dollar/blob/6e45c806a54869944550a6be8b102a1a751408a3/packages/contracts/src/dollar/core/UbiquityCreditToken.sol) and [CreditNft](https://github.com/ubiquity/ubiquity-dollar/blob/6e45c806a54869944550a6be8b102a1a751408a3/packages/contracts/src/dollar/core/CreditNft.sol) tokens. If those tokens are not going to be part of the production deploy then we need some mechanism for maintaining a USD peg.

The [Frax](https://frax.finance/) protocol could be of help.

Frax core contracts:
- [FRAX token](https://github.com/FraxFinance/frax-solidity/blob/fc9810d72c520d256965b81b6c9cc6aa95d07d9d/src/hardhat/contracts/Frax/Frax.sol): USD-pegged stablecoin
- [FXS token](https://github.com/FraxFinance/frax-solidity/blob/fc9810d72c520d256965b81b6c9cc6aa95d07d9d/src/hardhat/contracts/FXS/FXS.sol): seigniorage and governance token
- [FraxPool](https://github.com/FraxFinance/frax-solidity/blob/fc9810d72c520d256965b81b6c9cc6aa95d07d9d/src/hardhat/contracts/Frax/Pools/FraxPoolV3.sol): mints and redeems the FRAX stablecoin for collateral + helps maintaining the USD peg
- [AMO contracts](https://github.com/FraxFinance/frax-solidity/tree/fc9810d72c520d256965b81b6c9cc6aa95d07d9d/src/hardhat/contracts/Misc_AMOs/Lending_AMOs): use collateral from the [FraxPool](https://github.com/FraxFinance/frax-solidity/blob/fc9810d72c520d256965b81b6c9cc6aa95d07d9d/src/hardhat/contracts/Frax/Pools/FraxPoolV3.sol) to make yield
- [AMO minter](https://github.com/FraxFinance/frax-solidity/blob/fc9810d72c520d256965b81b6c9cc6aa95d07d9d/src/hardhat/contracts/Frax/FraxAMOMinter.sol): takes collateral from the [FraxPool](https://github.com/FraxFinance/frax-solidity/blob/fc9810d72c520d256965b81b6c9cc6aa95d07d9d/src/hardhat/contracts/Frax/Pools/FraxPoolV3.sol) and put it into one of the [AMOs](https://github.com/FraxFinance/frax-solidity/tree/fc9810d72c520d256965b81b6c9cc6aa95d07d9d/src/hardhat/contracts/Misc_AMOs/Lending_AMOs) (this is decided by the governance) 

Initially (in the [version 1](https://github.com/FraxFinance/frax-solidity/blob/fc9810d72c520d256965b81b6c9cc6aa95d07d9d/frax_whitepaper_v1.pdf)) the Frax protocol was meant to go through 3 stages:
1. 100% collateralized (ex: for 1 USDC you get 1 FRAX in the [FraxPool](https://github.com/FraxFinance/frax-solidity/blob/fc9810d72c520d256965b81b6c9cc6aa95d07d9d/src/hardhat/contracts/Frax/Pools/FraxPoolV3.sol))
2. Fractional (0-100% collateralized). The Frax protocol has a moving collateral ratio which is changed depending on the FRAX token price (i.e. USD peg). The more the collateral ratio is the less seigniorage tokens (i.e. FXS) you need to mint or redeem a single FRAX from the [FraxPool](https://github.com/FraxFinance/frax-solidity/blob/fc9810d72c520d256965b81b6c9cc6aa95d07d9d/src/hardhat/contracts/Frax/Pools/FraxPoolV3.sol). (ex: if collateral ratio is 90% then you need to transfer 0.9 USD in collateral + 0.1 USD in seignorage tokens (i.e. FXS) to the [FraxPool](https://github.com/FraxFinance/frax-solidity/blob/fc9810d72c520d256965b81b6c9cc6aa95d07d9d/src/hardhat/contracts/Frax/Pools/FraxPoolV3.sol) to mint 1 FRAX. On the contrary you will get 0.9 USD in collateral and 0.1 USD in FXS on redeeming).
3. Algorithmic (0% collateralized). On any mint/redeem of a single FRAX token you need only seigniorage tokens (i.e. FXS).

The Frax protocol has undergone 3 versions:
- `v1`: a single [FraxPool](https://github.com/FraxFinance/frax-solidity/blob/fc9810d72c520d256965b81b6c9cc6aa95d07d9d/src/hardhat/contracts/Frax/Pools/FraxPool.sol) responsible for maintaining the USD peg
- `v2`: [AMO contracts](https://github.com/FraxFinance/frax-solidity/tree/fc9810d72c520d256965b81b6c9cc6aa95d07d9d/src/hardhat/contracts/Misc_AMOs/Lending_AMOs) introduced
- `v3`: move to 100% collateral ([FIP](https://gov.frax.finance/t/fip-188-increase-cr-to-100/2147) + [voting](https://snapshot.org/#/frax.eth/proposal/0x6d43531f54de3d52ad439dcb6c56e3eae1b833fceb4d484dcb3b20be070ca25d))

We can see that `v3` of the protocol totally shifts the `v1`'s paradigm from fractional (i.e. partially collateralized) stablecoin to 100% collateralized stablecoin.

@pavlovcik Questions:
1. Will the Dollar token be fractional (i.e. partially collateralized) or fully collateralized? If it is fully collateralized then it is much simpler and faster to implement the [UbiquityPoolFacet](https://github.com/ubiquity/ubiquity-dollar/blob/6e45c806a54869944550a6be8b102a1a751408a3/packages/contracts/src/dollar/facets/UbiquityPoolFacet.sol) (i.e. forked [FraxPool](https://github.com/FraxFinance/frax-solidity/blob/fc9810d72c520d256965b81b6c9cc6aa95d07d9d/src/hardhat/contracts/Frax/Pools/FraxPoolV3.sol) responsible for USD peg).
2. If the Dollar token is going to be fully collateralized in its initial stage are there any plans to make it fractional (i.e. partially collateralized)?

**Comment by 0x4007** [comment]
Last time I put significant thought into this I came to the conclusion of 95% collateralized (LUSD) and 5% UBQ

This is because LUSD historically has fluctuated around 5%

For example, if LUSD is trading at 1.05, and our entire reserves are created at this moment ($1M) 

And then LUSD goes to 1.00

We owe $1M of collateral but we have only $952,380.9523809524 in the bank. (I'm not sure if my math is accurate but you get the idea)

The difference should hopefully come from UBQ value in case of a bank run scenario. 

---

In case there are liquidity issues with LUSD in the future, we can add sUSD and consider their economics to change the 5% UBQ ratio potentially. 

---

Lastly if we can make yield on the collateral, we can slowly close that 5% gap as well if we are inclined to.

**Reply by rndquu** [reply]
A few more questions:
1. The latest [FraxPoolV3](https://etherscan.io/address/0x2fE065e6FFEf9ac95ab39E5042744d695F560729) has the "pool ceiling" feature. It basically sets an upper limit on how much collateral the pool can accept. As far as I understand this is to prevent draining of large amounts of collateral in case of hacks. The latest [FraxPoolV3](https://etherscan.io/address/0x2fE065e6FFEf9ac95ab39E5042744d695F560729) has a 50k LUSD ceiling. What value should we set?
2. We should allow minting Dollar for collateral tokens only when Dollar price > 1.00, right?
3. We should allow redeeming collateral tokens for Dollar when Dollar price < 1.00, right?
4. The latest [FraxPoolV3](https://etherscan.io/address/0x2fE065e6FFEf9ac95ab39E5042744d695F560729) has the mint fee equal to 0.3%. What value should we set?
5. The latest [FraxPoolV3](https://etherscan.io/address/0x2fE065e6FFEf9ac95ab39E5042744d695F560729) has the redeem fee equal to 0.45%. What value should we set?
6. We should add a new "AMO minter" that can borrow collateral from the pool, right?
7. We're taking all prices (i.e. `LUSD/USD` and `Dollar/USD`) from Curve's TWAP, right?

**Reply by 0x4007** [reply]
> A few more questions:
> 
> 1. The latest [FraxPoolV3](https://etherscan.io/address/0x2fE065e6FFEf9ac95ab39E5042744d695F560729) has the "pool ceiling" feature. It basically sets an upper limit on how much collateral the pool can accept. As far as I understand this is to prevent draining of large amounts of collateral in case of hacks. The latest [FraxPoolV3](https://etherscan.io/address/0x2fE065e6FFEf9ac95ab39E5042744d695F560729) has a 50k LUSD ceiling. What value should we set?

Intuitively I feel that we should base our answer off of the total amount across all assets they allowed. For example if USDC is 250k, sUSD is 50k, LUSD is 50k for a total of 350k; then we should do 350k. I presume that this value is easy to update so I can put more thought into it later if necessary. 

> 2. We should allow minting Dollar for collateral tokens only when Dollar price > 1.00, right?
>
> 3. We should allow redeeming collateral tokens for Dollar when Dollar price < 1.00, right?

I hope that this is adjustable. The simple answer is yes, but intuitively I feel that there might be a benefit to having a small band. For example 0.995 and 1.005 being the trigger prices. I will need to dig up my old research around this, which is not conveniently accessible from my phone. 

> 4. The latest [FraxPoolV3](https://etherscan.io/address/0x2fE065e6FFEf9ac95ab39E5042744d695F560729) has the mint fee equal to 0.3%. What value should we set?
> 
> 5. The latest [FraxPoolV3](https://etherscan.io/address/0x2fE065e6FFEf9ac95ab39E5042744d695F560729) has the redeem fee equal to 0.45%. What value should we set?

Let's use their values as a default. I hope that these are adjustable. 

> 6. We should add a new "AMO minter" that can borrow collateral from the pool, right?

This is unclear to me. Can you provide me more context with what FRAX system this is?

> 7. We're taking all prices (i.e. `LUSD/USD` and `Dollar/USD`) from Curve's TWAP, right?

Yes

**Reply by rndquu** [reply]
> We should add a new "AMO minter" that can borrow collateral from the pool, right?
>> This is unclear to me. Can you provide me more context with what FRAX system this is?

I actually meant adding a new "AMO minter" **role** for contracts that can borrow collateral. The term "AMO minter" in the context of the Frax protocol is simply a smart contract that can take collateral from the FraxPool and put it to some protocol to make yield (compound, curve, etc...). Those "AMO minter" contracts can take collateral only by the governance. Those "AMO minter" contracts should also be explicitly added to a special whitelist of contracts which can borrow collateral. I just wanted to clarify that we need the same mechanic, and the answer seems to be "yes".

Regarding your other answers, all parameters are adjustable.
